File: examples/solutions/range_sum/ac.py
# Accepted: prefix sums, O(n + q). Uses sys.stdin for fast I/O.
import sys


def main() -> None:
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    q = int(data[idx]); idx += 1

    prefix = [0] * (n + 1)
    for i in range(1, n + 1):
        prefix[i] = prefix[i - 1] + int(data[idx]); idx += 1

    out = []
    for _ in range(q):
        left = int(data[idx]); idx += 1
        right = int(data[idx]); idx += 1
        out.append(str(prefix[right] - prefix[left - 1]))
    sys.stdout.write("\n".join(out) + ("\n" if out else ""))


main()

File: examples/solutions/range_sum/ac.cpp
// Accepted: prefix sums answer each query in O(1) -> O(n + q) overall.
#include <cstdio>
#include <vector>

int main() {
    int n, q;
    if (scanf("%d %d", &n, &q) != 2) return 0;
    std::vector<long long> prefix(n + 1, 0);
    for (int i = 1; i <= n; ++i) {
        long long x;
        scanf("%lld", &x);
        prefix[i] = prefix[i - 1] + x;
    }
    while (q--) {
        int l, r;
        scanf("%d %d", &l, &r);
        printf("%lld\n", prefix[r] - prefix[l - 1]);
    }
    return 0;
}
